libexec/rtld-elf/rtld_c18n.c

```diff
@@ -0,0 +1,700 @@
+	uint32_t x = key ^ (key >> 32);
+	x ^= x >> 16;
+	x *= 0x21f0aaadU;
+	x ^= x >> 15;
+	x *= 0xd35a2d97U;
+	x ^= x >> 15;
+	return (x);
+}
+
+static slot_idx_t
+nextSlot(slot_idx_t slot, uint32_t hash, int exp)
+{
+	uint32_t mask = (1 << exp) - 1;
+	uint32_t step = (hash >> (32 - exp)) | 1;
+	return ((slot + step) & mask);
+}
+
+static void
+resizeTable(int exp)
+{
+	struct tramp_map_kv *map;
+	struct tramp_data *data;
+	ptraddr_t key;
+	uint32_t hash;
+	slot_idx_t size, slot;
+
+	assert(0 < exp && exp < 32);
+
+	free(tramp_table.map);
+	data = tramp_data_expand(exp);
+	map = tramp_map_new(exp);
+	size = atomic_load_explicit(&tramp_table.size, memory_order_relaxed);
+
+	for (slot_idx_t idx = 0; idx < size; ++idx) {
+		key = (ptraddr_t)data[idx].target;
+		hash = pointerHash(key);
+		slot = hash;
+		do
+			slot = nextSlot(slot, hash, exp);
+		while (atomic_load_explicit(
+		    &map[slot].key, memory_order_relaxed) != 0);
+		atomic_store_explicit(&map[slot].key, key,
+		    memory_order_relaxed);
+		atomic_store_explicit(&map[slot].index, idx,
+		    memory_order_relaxed);
+	}
+
+	tramp_table.data = data;
+	tramp_table.map = map;
+	tramp_table.exp = exp;
+}
+
+void
+tramp_hook(int event, void *target, const Obj_Entry *obj, const Elf_Sym *def,
+    void *link, void *rcsp)
+{
+	Elf64_Word sym_num;
+	const char *sym;
+	const char *callee;
+
+	compart_id_t caller_id;
+	const char *caller;
+
+	sym_num = def == NULL ? 0 : def->st_name;
+	sym = def == NULL ? "<unknown>" : strtab_value(obj, def->st_name);
+	callee = comparts.data[obj->compart_id]->name;
+
+	if ((cheri_getperm(link) & CHERI_PERM_EXECUTIVE) == 0)
+		caller_id = ((uintptr_t *)
+		    cheri_setaddress(rcsp, cheri_gettop(rcsp)))[-2];
+	else
+		caller_id = 0;
+	caller = comparts.data[caller_id]->name;
+
+	if (ld_compartment_utrace != NULL)
+		ld_utrace_log(event,
+			target, NULL, sym_num, 0, callee, sym, caller);
+	if (ld_compartment_overhead != NULL)
+		getpid();
+}
+
+static void *
+tramp_pgs_append(const struct tramp_data *data)
+{
+	size_t len;
+	/* A capability-aligned buffer large enough to hold a trampoline */
+	tramp tmp[48];
+	tramp *tramp, *entry = tmp;
+	struct tramp_pg *pg;
+
+	/* Fill a temporary buffer with the trampoline and obtain its length */
+	len = tramp_compile(&entry, data);
+
+	pg = atomic_load_explicit(&tramp_pgs.head, memory_order_acquire);
+	tramp = tramp_pg_push(pg, len);
+	if (!cheri_gettag(tramp)) {
+		while (atomic_flag_test_and_set_explicit(&tramp_pgs.lock,
+		    memory_order_acquire));
+		pg = atomic_load_explicit(&tramp_pgs.head,
+		    memory_order_relaxed);
+		tramp = tramp_pg_push(pg, len);
+		if (!cheri_gettag(tramp)) {
+			pg = tramp_pg_new(pg);
+			tramp = tramp_pg_push(pg, len);
+			atomic_store_explicit(&tramp_pgs.head, pg,
+			    memory_order_release);
+		}
+		atomic_flag_clear_explicit(&tramp_pgs.lock,
+		    memory_order_release);
+	}
+	assert(cheri_gettag(tramp));
+
+	entry = tramp + (entry - tmp);
+	tramp = mempcpy(tramp, tmp, len);
```

Comment:
Overwriting tramp is unnecessarily confusing, just use a normal memcpy and add len in your cheri_copyaddress call